### PR TITLE
ffi module cleanup: context.h to frameobject.h

### PR DESCRIPTION
**Pull request by nw0**
Partial fix for #1289. This set seems relatively more straightforward, but touches a lot of files.

### CODE REVIEW

**Review by davidhewitt — COMMENTED** [review]
Thanks again for continuing on with this, this is looking excellent! I see a few TODO comments so I'll give this another look through once this is resolved.

Regarding the CHANGELOG - we usually make the sections `Added`, `Changed`, `Removed`, `Fixed` in that order, so if you're adding a new section would be great to make it fit that ordering.

File: src/ffi/descrobject.rs [hunk 1]
```diff
@@ -65,10 +58,40 @@ extern "C" {
...
     pub fn PyDictProxy_New(arg1: *mut PyObject) -> *mut PyObject;
     pub fn PyWrapper_New(arg1: *mut PyObject, arg2: *mut PyObject) -> *mut PyObject;
 
     #[cfg_attr(PyPy, link_name = "PyPyProperty_Type")]
     pub static mut PyProperty_Type: PyTypeObject;
 }
+
+/// Helper initial value of [`PyGetSetDef`] for a Python class.
+///
+/// Not present in `cpython/Include/descrobject`.
```
[the remark this answers is not in the document]
Comment (nw0) [reply src/ffi/descrobject.rs:73]: Yes, I don't find any obvious origin in cpython. If `ffi` is strictly for imports, we should probably move this, but it's up to you.

[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/ffi/descrobject.rs:73]: Yeah, I think this (and `PyGetSetDef_DICT`) should probably be moved to just be private variables in `pyclass.rs`.

For now please just mark these deprecated; I'll move them in a future breaking release.

File: src/ffi/dictobject.rs [hunk 2]
```diff
@@ -130,6 +64,60 @@ extern "C" {
...
+    pub static mut PyDictRevIterValue_Type: PyTypeObject;
+    #[cfg(Py_3_8)]
+    pub static mut PyDictRevIterItem_Type: PyTypeObject;
 }
+
+#[repr(C)]
+#[derive(Debug)]
+// TODO: move to cpython/dictobject.rs
+// Not moved because dict.rs uses PyDictObject extensively.
+pub struct PyDictObject {
```
Comment (nw0) [inline src/ffi/dictobject.rs:114]: @davidhewitt re "TODO"s:

`PyDictObject` is a (non-limited) CPython API, but used by dict.rs unconditionally. I've therefore left it in ffi/dictobject.rs, even though the definition is from CPython.

I'm not sure removing this dependency can fall within the scope of #1289. Do you have any suggestions as to how to deal with it? My inclination is to leave it in place for now and spawn another issue to address it if necessary.

File: src/ffi/dictobject.rs [hunk 3]
```diff
@@ -130,6 +64,60 @@ extern "C" {
...
+    pub ob_base: PyObject,
+    pub ma_used: Py_ssize_t,
+    pub ma_version_tag: u64,
+    pub ma_keys: *mut PyDictKeysObject,
+    pub ma_values: *mut *mut PyObject,
+}
+
+// TODO: move to cpython/dictobject.rs
+opaque_struct!(PyDictKeysObject);
```
Comment (nw0) [inline src/ffi/dictobject.rs:123]: Again, not moved because `PyDictObject` depends on this.

File: src/ffi/floatobject.rs [hunk 4]
```diff
@@ -1,6 +1,8 @@
 use crate::ffi::object::*;
 use std::os::raw::{c_double, c_int};
 
+// TODO: mark non-limited
+// currently used by types/floatob.rs
 #[repr(C)]
 pub struct PyFloatObject {
```
Comment (nw0) [inline src/ffi/floatobject.rs:7]: Similar issue to `PyDictObject`. This is not a limited API, but is used unconditionally to define `PyFloat` in floatob.rs

**Review by nw0 — COMMENTED** [review]
> I see a few TODO comments so I'll give this another look through once this is resolved.

I've annotated these. Basically, these can't be moved as-is (or mark them non-limited), because we have code in dict.rs and floatob.rs (`py_native_type` invocations) that unconditionally uses these types. Could you advise on the appropriate action here?

**Comment by davidhewitt** [comment]
I think the right solution for `PyDictObject` and `PyFloatObject` is to use `opaque_struct!` with the limited API for now. You could also mark these with TODO to remove the `opaque_struct!` implementations later; we have to revisit this in the future anyway to enable subclassing builtin types with the `abi3` feature:

https://github.com/PyO3/pyo3/blob/3f093d9e59df0140f6c3a0abbb6acaa439687b7f/src/pyclass.rs#L44

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,10 +6,21 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
 
 ## [Unreleased]
+### Changed
+- Deprecate FFI definition `PyCoro_Check` and `PyAsyncGen_Check` in favor of `PyCoro_CheckExact` and `PyAsyncGen_CheckExact` respectively, as these are the correct name in the Python headers. [#1341](https://github.com/PyO3/pyo3/pull/1341)
+- Deprecate FFI definition `PyCoroWrapper_Check` and `PyGetSetDef_DICT` which have never been defined in the Python headers. [#1341](https://github.com/PyO3/pyo3/pull/1341)
+- Deprecate FFI definition `PyImport_Cleanup`, which was removed in 3.9 and previously marked "for internal use only". [#1341](https://github.com/PyO3/pyo3/pull/1341)
+- Deprecate FFI definition `PyOS_InitInterrupts`, which was removed in 3.10 and previously undocumented. [#1341](https://github.com/PyO3/pyo3/pull/1341)
+- Deprecate FFI definition `PyOS_AfterFork`; introduce `PyOS_BeforeFork`, `PyOS_AfterFork_Parent`, `PyOS_AfterFork_Child` when building for Python 3.7. [#1341](https://github.com/PyO3/pyo3/pull/1341)
+- Deprecate FFI definitions `PyEval_CallObjectWithKeywords`, `PyEval_CallObject`, `PyEval_CallFunction`, `PyEval_CallMethod` when building for Python 3.9. [#1338](https://github.com/PyO3/pyo3/pull/1338)
```
Comment (davidhewitt) [inline CHANGELOG.md:15]: Usually within the subsections we order by the individual PRs, so would prefer if this one is at the top of `Changed` please!

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,10 +6,21 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Changed
+- Deprecate FFI definition `PyCoro_Check` and `PyAsyncGen_Check` in favor of `PyCoro_CheckExact` and `PyAsyncGen_CheckExact` respectively, as these are the correct name in the Python headers. [#1341](https://github.com/PyO3/pyo3/pull/1341)
```
Comment (davidhewitt) [inline CHANGELOG.md:10]: 👍 changelog message looks good, but I don't see the `#[deprecated]` attribute in `genobject.rs`?

```suggestion
- Deprecate FFI definitions `PyCoro_Check` and `PyAsyncGen_Check` in favor of `PyCoro_CheckExact` and `PyAsyncGen_CheckExact` respectively (the former are not part of the Python API). [#1341](https://github.com/PyO3/pyo3/pull/1341)
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,10 +6,21 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+- Deprecate FFI definition `PyCoroWrapper_Check` and `PyGetSetDef_DICT` which have never been defined in the Python headers. [#1341](https://github.com/PyO3/pyo3/pull/1341)
```
Comment (davidhewitt) [inline CHANGELOG.md:11]: Maybe also `PyGetSetDef_INIT`?

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,10 +6,21 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+- Deprecate FFI definition `PyImport_Cleanup`, which was removed in 3.9 and previously marked "for internal use only". [#1341](https://github.com/PyO3/pyo3/pull/1341)
+- Deprecate FFI definition `PyOS_InitInterrupts`, which was removed in 3.10 and previously undocumented. [#1341](https://github.com/PyO3/pyo3/pull/1341)
```
Comment (davidhewitt) [inline CHANGELOG.md:13]: I think the detail here is slightly too much for a CHANGELOG (users can find out more from the PR link)

```suggestion
- Deprecate FFI definition `PyImport_Cleanup`  (removed in Python 3.9). [#1341](https://github.com/PyO3/pyo3/pull/1341)
- Deprecate FFI definition `PyOS_InitInterrupts` (removed in Python 3.10). [#1341](https://github.com/PyO3/pyo3/pull/1341)
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,10 +6,21 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+- Deprecate FFI definition `PyOS_AfterFork`; introduce `PyOS_BeforeFork`, `PyOS_AfterFork_Parent`, `PyOS_AfterFork_Child` when building for Python 3.7. [#1341](https://github.com/PyO3/pyo3/pull/1341)
```
Comment (davidhewitt) [inline CHANGELOG.md:14]: The new methods should be in the `Added` section please :)

File: src/ffi/context.rs [hunk 10]
```diff
@@ -34,3 +24,22 @@ pub unsafe fn PyContextVar_CheckExact(op: *mut PyObject) -> c_int {
...
+
+    pub fn PyContextVar_New(name: *const c_char, def: *mut PyObject) -> *mut PyObject;
+    pub fn PyContextVar_Get(
+        var: *mut PyObject,
+        default_value: *mut PyObject,
+        value: *mut *mut PyObject,
+    ) -> c_int;
+    pub fn PyContextVar_Set(var: *mut PyObject, value: *mut PyObject) -> *mut PyObject;
+    pub fn PyContextVar_Reset(var: *mut PyObject, token: *mut PyObject) -> c_int;
+// skipped non-limited _PyContext_NewHamtForTests
```
Comment (davidhewitt) [inline src/ffi/context.rs:44]:
```suggestion
    // skipped non-limited _PyContext_NewHamtForTests
```

File: src/ffi/cpython/dictobject.rs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    // skipped _PyDict_SetItemId
+    // skipped _PyDict_DelItemId
+    // skipped _PyDict_DebugMallocStats
+    // skipped _PyObjectDict_SetItem
+    // skipped _PyDict_LoadGlobal
+    // skipped _PyDict_GetItemHint
+    // skipped _PyDictViewObject
+    // skipped _PyDictView_New
+    // skipped _PyDictView_Intersect
+    // FIXME: PyDict_Contains is defined in dictobject.c
```
Comment (davidhewitt) [inline src/ffi/cpython/dictobject.rs:61]: huh?

File: src/ffi/cpython/frameobject.rs [hunk 12]
```diff
@@ -71,6 +80,9 @@ extern "C" {
     pub fn PyFrame_FastToLocalsWithError(f: *mut PyFrameObject) -> c_int;
     pub fn PyFrame_FastToLocals(f: *mut PyFrameObject);
 
+    // skipped _PyFrame_DebugMallocStats
+    // skipped PyFrame_GetBack
+
+    // FIXME: PyFrame_ClearFreeList is defined in frameobject.c
```
Comment (davidhewitt) [inline src/ffi/cpython/frameobject.rs:86]: Similarly not sure what the FIXME is asking us to do?

File: src/ffi/fileobject.rs [hunk 13]
```diff
@@ -14,19 +14,22 @@ extern "C" {
...
+    #[cfg_attr(PyPy, link_name = "PyPyFile_AsFileDescriptor")]
+    pub fn PyObject_AsFileDescriptor(arg1: *mut PyObject) -> c_int;
 }
 
 #[cfg_attr(windows, link(name = "pythonXY"))]
 extern "C" {
     pub static mut Py_FileSystemDefaultEncoding: *const c_char;
     pub static mut Py_FileSystemDefaultEncodeErrors: *const c_char;
     pub static mut Py_HasFileSystemDefaultEncoding: c_int;
+// skipped Python 3.7 / ex-non-limited Py_UTF8Mode
```
Comment (davidhewitt) [inline src/ffi/fileobject.rs:32]:
```suggestion
    // skipped Python 3.7 / ex-non-limited Py_UTF8Mode
```

**Review by davidhewitt — COMMENTED** [review]
😍 thank you again, so pleasing to see these files getting some proper treatment!

I added a few minor suggestions for the CHANGELOG and also some minor nits on indentation. Otherwise, this one's looking close to merging?

File: src/ffi/context.rs [hunk 14]
```diff
@@ -3,21 +3,11 @@ use std::os::raw::{c_char, c_int};
...
-    pub fn PyContext_Exit(ctx: *mut PyObject) -> c_int;
-    pub fn PyContextVar_New(name: *const c_char, def: *mut PyObject) -> *mut PyObject;
-    pub fn PyContextVar_Get(
-        var: *mut PyObject,
-        default_value: *mut PyObject,
-        value: *mut *mut PyObject,
-    ) -> c_int;
-    pub fn PyContextVar_Set(var: *mut PyObject, value: *mut PyObject) -> *mut PyObject;
-    pub fn PyContextVar_Reset(var: *mut PyObject, token: *mut PyObject) -> c_int;
+// skipped non-limited opaque PyContextToken
```
[the remark this answers is not in the document]
Comment (nw0) [reply src/ffi/context.rs:10]: This fails `cargo fmt -- --check` -- it's really odd. It's as if the comment gets "attached" to the next element, which is the closing brace. Ditto for the other indentation changes.

File: CHANGELOG.md (hunk 6 above)
Comment (nw0) [reply CHANGELOG.md:10]: Oops, I accidentally staged this. Removing.

File: src/ffi/context.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/ffi/context.rs:10]: Ah that's annoying but makes sense!

File: src/ffi/cpython/dictobject.rs (hunk 11 above)
Comment (nw0) [reply src/ffi/cpython/dictobject.rs:61]: Oops -- I was in the middle of tracking this down, will update. Basically it's removed in 3.10, but not in the Python changelog (yet).

**Comment by nw0** [comment]
Sorry about the mess. I've eyeballed the PR diff this time and think it's ready for review, if tests pass.

**Review by kngwyu — APPROVED** [review]
Thanks!

**Review by davidhewitt — APPROVED** [review]
Thank you again for your continued work on this cleanup! 💯

**Pull request merged**